File: common/src/main/java/com/imsproject/common/dataAccess/abstracts/SQLExecutor.java
package com.imsproject.common.dataAccess.abstracts;


import com.imsproject.common.dataAccess.OfflineResultSet;

import java.sql.SQLException;

public interface SQLExecutor {

    /**
     * Begins a transaction
     * @throws SQLException if an error occurs while starting the transaction
     */
    void beginTransaction() throws SQLException;

    /**
     * Commits the transaction
     * @throws SQLException if an error occurs while committing the transaction
     */
    void commit() throws SQLException;

    /**
     * Rolls back the transaction
     * @throws SQLException if an error occurs while rolling back the transaction
     */
    void rollback() throws SQLException;

    /**
     * Executes a read query by using {@link java.sql.PreparedStatement}
     * @param query An sql un-prepared query. Use '?' for parameters
     * @param params The parameters to be inserted into the query in order (will replace '?' in the query)
     * @return An {@link OfflineResultSet} containing the result of the query
     * @throws SQLException if an error occurs while executing the query
     */
    OfflineResultSet executeRead(String query, Object ... params) throws SQLException;

    /**
     * Executes a write query by using {@link java.sql.PreparedStatement}
     * @param query An sql un-prepared query. Use '?' for parameters
     * @param params The parameters to be inserted into the query in order (will replace '?' in the query)
     * @return The number of rows affected by the query
     * @throws SQLException if an error occurs while executing the query
     */
    int executeWrite(String query, Object ... params) throws SQLException;
}

File: common/src/main/java/com/imsproject/common/dataAccess/abstracts/SQLExecutor.kt
package com.imsproject.common.dataAccess.abstracts

import com.imsproject.common.dataAccess.OfflineResultSet
import java.sql.SQLException


interface SQLExecutor {
    /**
     * Begins a transaction
     * @throws SQLException if an error occurs while starting the transaction
     * @return transaction id
     */
    @Throws(SQLException::class)
    fun beginTransaction() : String

    /**
     * Commits the transaction
     * @throws SQLException if an error occurs while committing the transaction
     */
    @Throws(SQLException::class)
    fun commit(transactionId: String)

    /**
     * Rolls back the transaction
     * @throws SQLException if an error occurs while rolling back the transaction
     */
    @Throws(SQLException::class)
    fun rollback(transactionId: String)

    /**
     * Executes a read query by using [java.sql.PreparedStatement]
     * @param query A sql prepared query. Use '?' for parameters
     * @param params The parameters to be used in the query in order
     * @return An [OfflineResultSet] containing the result of the query
     * @throws SQLException if an error occurs while executing the query
     */
    @Throws(SQLException::class)
    fun executeRead(query: String, params: Array<out Any?> = emptyArray(), transactionId: String? = null): OfflineResultSet

    /**
     * Executes a generic write query by using [java.sql.PreparedStatement].
     * Can be used for either insert, update, delete
     * @param query A sql prepared query. Use '?' for parameters
     * @param params The parameters to be used in the query in order
     * @return The number of rows affected by the query
     * @throws SQLException if an error occurs while executing the query
     */
    @Throws(SQLException::class)
    fun executeWrite(query: String, params: Array<out Any?> = emptyArray(), transactionId: String? = null): Int

    /**
     * Executes an insert query by using [java.sql.PreparedStatement]
     * @param query A sql prepared query. Use '?' for parameters
     * @param params The parameters to be used in the query in order
     * @return A [OfflineResultSet] containing the generated keys
     */
    @Throws(SQLException::class)
    fun executeInsert(query: String, params: Array<out Any?> = emptyArray(), transactionId: String? = null): OfflineResultSet

    /**
     * Executes an update or delete query by using [java.sql.PreparedStatement]
     * @param query A sql prepared query. Use '?' for parameters
     * @param params The parameters to be used in the query in order
     * @return The number of rows affected by the query
     * @throws SQLException if an error occurs while executing the query
     */
    @Throws(SQLException::class)
    fun executeUpdateDelete(query: String, params: Array<out Any?> = emptyArray(), transactionId: String? = null): Int

    /**
     * Executes a bulk insert query by using [java.sql.PreparedStatement]
     * @param query A sql prepared query. Use '?' for parameters
     * @param params The parameters to be used in the query in order
     * @return The number of rows affected by the query
     * @throws SQLException if an error occurs while executing the query
     */
    @Throws(SQLException::class)
    fun executeBulkInsert(query: String, params: List<Array<out Any?>> = emptyList(), transactionId: String? = null): Int
}
